### DISCUSSION TITLE
visionOS support

### DISCUSSION THREAD

**Discussion by wtholliday** [microsoft/onnxruntime#19313] [Mobile Q&A] [answered]
I'm working on visionOS support here: https://github.com/AudioKit/onnxruntime/tree/visionos

Here's a quick link to compare my branch: https://github.com/microsoft/onnxruntime/compare/main...AudioKit:onnxruntime:visionos

My build command is:
```
python3 tools/ci_build/github/apple/build_apple_framework.py tools/ci_build/github/apple/default_mobile_visionos_framework_build_settings.json --config Release
```

Error I'm getting is:
```
onnxruntime/core/providers/coreml/builders/coreml_spec.h:12:10: fatal error: 'coreml/Model.pb.h' file not found
#include "coreml/Model.pb.h"
         ^~~~~~~~~~~~~~~~~~~
```

Any help greatly appreciated :)

**Comment by yufenglee** [comment]
@skottmckay, could you please help on this?

**Comment by skottmckay** [comment]
That file would usually be generated by the build. Not sure what values cmake has for a visionOS build and whether the conditions at the start of this block need updating to support it. 

https://github.com/microsoft/onnxruntime/blob/465540d29b87bcbe5b4bc1d7ff4e7fb082693746/cmake/onnxruntime_providers_coreml.cmake#L11-L37

**Reply by wtholliday** [reply]
Thanks! I changed some of those conditions and it builds know (branch is updated). lmk if you'd like me to make it into a proper PR.

**Reply by snnn** [reply]
Should be "visionOS".
See: https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-ios-tvos-visionos-or-watchos

**Reply by skottmckay** [reply]
[This](https://github.com/microsoft/onnxruntime/pull/19347) PR is indirectly reworking the CoreML build setup a little to build the protobuf files on all platforms, so that specific check will go away. If there are other places that need to be aware of 'visionOS' a PR for those would be great.

**Comment by Ciyou** [comment]
I tried @wtholliday's fork and it does works with minimal build! But if I remove the `--minimal_build` option the build failed with below error:

```
ld: Undefined symbols:
  Provider_GetHost, referenced from:
      <initial-undefines>
  Provider_SetHost, referenced from:
      <initial-undefines>
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

My build command is:

```
['/opt/homebrew/opt/python@3.11/bin/python3.11', '/Users/xxx/Developer/onnxruntime/tools/ci_build/build.py', '--parallel', '--use_xcode', '--build_apple_framework', '--use_coreml', '--skip_tests', '--cmake_extra_defines=onnxruntime_BUILD_UNIT_TESTS=OFF', '--ios', '--apple_deploy_target=1.0', '--config=Release', '--apple_sysroot=xros', '--osx_arch=arm64', '--build_dir=/Users/ciyou/Developer/onnxruntime/build/apple_framework/intermediates/xros_arm64']
```

Wondering which piece is missing for the standard build with xros SDK. 🤔

**Comment by skottmckay** [comment, marked as answer]
Add "visionOS" here: https://github.com/microsoft/onnxruntime/blob/58f4921686bf0a5b0442fb6df92d1b1972a118cc/cmake/onnxruntime_providers_cpu.cmake#L226
